pkg/controller/etcdproxy/certs.go

```diff
@@ -7,198 +7,350 @@ import (
+// Expired CA certificates from the bundle are removed in this phase.
+// * Generates new Server certificate/key pair using the newly generated CA certificate and update Secret in the controller namespace with new pair.
+// * "Restarts" etcd-proxy to pick-up new changes.
+// TODO: Implement "restarting" etcd-proxy. This is going to be done using rolling updates.
+func (c *EtcdProxyController) ensureServerCertificates(etcdstorage *etcdstoragev1alpha1.EtcdStorage) error {
+	serverSecret, err := c.getSeverCertSecret(etcdstorage)
 	if err != nil {
 		return err
 	}
-	serverSignerCert, _, err := serverSigner.GetPEMBytes()
+
+	// Check is annotation containing expiry date present and valid. If it is valid, we're skipping this iteration.
+	if expiry, ok := serverSecret.Annotations[ProxyCertificateExpiryAnnotation]; ok {
+		certExpiry, err := time.Parse(time.RFC3339, expiry)
+		if err != nil {
+			return err
+		}
+		if certExpiry.After(time.Now()) {
+			return nil
+		}
+	}
+
+	// Generate new Serving CA certificate and Server certificate/key pair.
+	servingCA, serverCert, err := c.generateServerBundle(etcdstorage)
 	if err != nil {
 		return err
 	}
 
-	// Generate server certificate/key pair.
-	serverBundle, err := serverSigner.NewServerCertificate(pkix.Name{CommonName: serviceUrl},
-		[]string{serviceUrl}, r.Int63n(100000), currentTime)
+	// Write new Server certificate/key pair to the Secret in the controller namespace.
+	serverCertBytes, serverKeyBytes, err := serverCert.GetPEMBytes()
 	if err != nil {
 		return err
 	}
-	serverCertBytes, serverKeyBytes, err := serverBundle.GetPEMBytes()
+
+	serverSecret.Annotations = map[string]string{
+		ProxyCertificateExpiryAnnotation: serverCert.Certificates[0].NotAfter.Format(time.RFC3339),
+	}
+	serverSecret.Data = map[string][]byte{
+		"tls.crt": serverCertBytes,
+		"tls.key": serverKeyBytes,
+	}
+	err = ensureSecret(c.kubeclientset, serverSecret)
 	if err != nil {
 		return err
 	}
 
-	// Write pairs to appropriate ConfigMaps and Secrets.
+	// Append new Serving CA certificate to the bundle in all ConfigMaps defined by EtcdStorage Spec.
 	var errs []error
-	for _, configMap := range etcdstorage.Spec.CACertConfigMaps {
-		err = c.ensureServingCABundle(etcdstorage, configMap, serverSignerCert, serverSigner.Certificates[0].NotAfter)
+	for _, cm := range etcdstorage.Spec.CACertConfigMaps {
+		// Get CA bundle from the ConfigMap, check does it already have certificates in the bundle, append new one to it,
+		// and filter expired certificates.
+		configMap, err := c.getServingCABundleConfigMap(etcdstorage, cm)
+		if err != nil {
+			return err
+		}
+		var ca *certs.Certificate
+		if oldCABytes, ok := configMap.Data["serving-ca.crt"]; ok {
+			ca, err = certs.ParseCertificateBytes([]byte(oldCABytes), nil)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+		}
+		// TODO: This is *very* ugly.
```

Comment:
why?

Reply:
> why?

Same question.  Matches my expectations.
 1. get current authoritative data (above)
 2. create new authoritative data if required (above)
 3. iterate over downstream consumers of data
 4. for each one, get it's current content if applicable
 5. add new authoritative data if required (are you missing this step)?
 6. remove unnecessary data
 7. update consumer

what else would you do?